lib/features/base/before_unload_manager.dart

```diff
@@ -0,0 +1,17 @@
+typedef BeforeUnloadListener = Future<void> Function();
+
+class BeforeUnloadManager {
+  static final BeforeUnloadManager _instance = BeforeUnloadManager._();
+  factory BeforeUnloadManager() => _instance;
+  BeforeUnloadManager._();
+
+  final _listeners = <BeforeUnloadListener>[];
+
+  void addListener(BeforeUnloadListener listener) {
+    _listeners.add(listener);
+  }
+
+  Future<void> executeBeforeUnloadListeners() async {
```

Comment:
we should add some tests for this functions

Reply:
Test added